/*
 * SPDX-License-Identifier: MIT
 * SPDX-FileCopyrightText: Copyright (c) 2021 Jason Skuby (mytechtoybox.com)
 */

#ifndef DISPLAY_H_
#define DISPLAY_H_

#include <map>
#include <string>
#include <hardware/i2c.h>
#include "BoardConfig.h"
#include "GPGFX.h"
#include "GPGFX_UI.h"
#include "GPGFX_UI_types.h"
#include "GPGFX_UI_screens.h"
#include "GPGFX_UI_widgets.h"
#include "gpaddon.h"
#include "gamepad.h"
#include "storagemanager.h"
#include "peripheralmanager.h"
#include "peripheral_i2c.h"
#include "peripheral_spi.h"

#ifndef HAS_I2C_DISPLAY
#define HAS_I2C_DISPLAY -1
#endif

#ifndef DISPLAY_I2C_ADDR
#define DISPLAY_I2C_ADDR 0x3C
#endif

#ifndef I2C_SDA_PIN
#define I2C_SDA_PIN -1
#endif

#ifndef I2C_SCL_PIN
#define I2C_SCL_PIN -1
#endif

#ifndef DISPLAY_I2C_BLOCK
#define DISPLAY_I2C_BLOCK i2c0
#endif

#ifndef I2C_SPEED
#define I2C_SPEED 800000
#endif

#ifndef DISPLAY_SIZE
#define DISPLAY_SIZE GPGFX_DisplaySize::SIZE_128x64
#endif

#ifndef DISPLAY_FLIP
#define DISPLAY_FLIP 0
#endif

#ifndef DISPLAY_INVERT
#define DISPLAY_INVERT 0
#endif

#ifndef DISPLAY_USEWIRE
#define DISPLAY_USEWIRE 1
#endif

#ifndef DISPLAY_SAVER_TIMEOUT
#define DISPLAY_SAVER_TIMEOUT 0
#endif

#ifndef BUTTON_LAYOUT
#define BUTTON_LAYOUT BUTTON_LAYOUT_STICK
#endif

#ifndef BUTTON_LAYOUT_RIGHT
#define BUTTON_LAYOUT_RIGHT BUTTON_LAYOUT_VEWLIX
#endif

#ifndef SPLASH_MODE
#define SPLASH_MODE SPLASH_MODE_NONE
#endif

#ifndef SPLASH_CHOICE
#define SPLASH_CHOICE SPLASH_CHOICE_MAIN
#endif

#ifndef SPLASH_DURATION
#define SPLASH_DURATION 7500 // Duration in milliseconds
#endif

#ifndef DISPLAY_TURN_OFF_WHEN_SUSPENDED
#define DISPLAY_TURN_OFF_WHEN_SUSPENDED 0
#endif

#ifndef INPUT_HISTORY_ENABLED
#define INPUT_HISTORY_ENABLED 0
#endif

#ifndef INPUT_HISTORY_LENGTH
#define INPUT_HISTORY_LENGTH 21
#endif

#ifndef INPUT_HISTORY_COL
#define INPUT_HISTORY_COL 0
#endif

#ifndef INPUT_HISTORY_ROW
#define INPUT_HISTORY_ROW 7
#endif

#ifndef DEFAULT_SPLASH
#define DEFAULT_SPLASH \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x01,0xFF,0x80,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x7F,0xC0,0x00,0x00,0x00,0x00, \
	0x07,0xFF,0xE0,0x3F,0xE7,0xFE,0x1F,0xE0,0xFF,0x01,0xFF,0xF0,0x00,0x00,0x00,0x00, \
	0x0F,0xFF,0xF8,0x7F,0xE7,0xFE,0x1F,0xE0,0xFF,0x07,0xFF,0xF8,0x00,0x0F,0xF0,0x00, \
	0x1F,0xFF,0xFC,0x7F,0xE7,0xFE,0x1F,0xF0,0xFF,0x0F,0xFF,0xFC,0x00,0x0F,0xF0,0x00, \
	0x1F,0xFF,0xFC,0x7F,0xE7,0xFE,0x1F,0xF0,0xFF,0x0F,0xFF,0xFE,0x00,0x1F,0xF0,0x00, \
	0x3F,0xFF,0xFC,0x7F,0xE7,0xFE,0x1F,0xF0,0xFF,0x1F,0xFF,0xFE,0x00,0x1F,0xF8,0x00, \
	0x3F,0xF7,0xFC,0x7F,0xE7,0xFE,0x1F,0xF8,0xFF,0x1F,0xFB,0xFF,0x00,0x1F,0xF8,0x00, \
	0x3F,0xE7,0xFE,0x3F,0xE7,0xFE,0x1F,0xF8,0xFF,0x1F,0xF3,0xFF,0x00,0x3F,0xF8,0x00, \
	0x3F,0xE7,0xFE,0x7F,0xE7,0xFE,0x1F,0xF8,0xFF,0x1F,0xF1,0xFF,0x00,0x3F,0xFC,0x00, \
	0x3F,0xE7,0xFE,0x7F,0xE7,0xFE,0x1F,0xF8,0xFF,0x3F,0xF1,0xFF,0x00,0x7F,0xFC,0x00, \
	0x3F,0xE7,0xFE,0x7F,0xE7,0xFE,0x1F,0xFC,0xFF,0x3F,0xF1,0xFF,0x00,0x7F,0xFE,0x00, \
	0x3F,0xE7,0xFE,0x7F,0xE7,0xFE,0x1F,0xFC,0xFF,0x3F,0xF1,0xFF,0x00,0x7F,0xFE,0x00, \
	0x3F,0xF0,0x00,0x3F,0xE7,0xFE,0x1F,0xFC,0xFF,0x3F,0xF1,0xFF,0x00,0xFF,0xFE,0x00, \
	0x3F,0xF8,0x00,0x7F,0xE7,0xFE,0x1F,0xFE,0xFF,0x3F,0xF1,0xFF,0x00,0xFF,0xFF,0x00, \
	0x3F,0xFE,0x00,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF0,0x00,0x00,0xFF,0xFF,0x00, \
	0x1F,0xFF,0x00,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF0,0x00,0x01,0xFF,0xFF,0x00, \
	0x1F,0xFF,0xC0,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF0,0x00,0x01,0xFF,0xFF,0x80, \
	0x0F,0xFF,0xE0,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF7,0xFF,0x01,0xFE,0xFF,0x80, \
	0x07,0xFF,0xF0,0x3F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF7,0xFF,0x03,0xFE,0xFF,0x80, \
	0x03,0xFF,0xF8,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF7,0xFF,0x03,0xFE,0x7F,0xC0, \
	0x00,0xFF,0xFC,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF7,0xFF,0x03,0xFC,0x7F,0xC0, \
	0x00,0x7F,0xFE,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF7,0xFF,0x07,0xFC,0x7F,0xC0, \
	0x00,0x1F,0xFE,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF1,0xFF,0x07,0xFC,0x3F,0xE0, \
	0x00,0x0F,0xFE,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF1,0xFF,0x07,0xF8,0x3F,0xE0, \
	0x3F,0xE7,0xFE,0x7F,0xE7,0xFE,0x1F,0xFF,0xFF,0x3F,0xF1,0xFF,0x0F,0xF8,0x3F,0xE0, \
	0x3F,0xE3,0xFE,0x7F,0xE7,0xFE,0x1F,0xEF,0xFF,0x3F,0xF1,0xFF,0x0F,0xF8,0x1F,0xF0, \
	0x3F,0xE3,0xFE,0x7F,0xE7,0xFE,0x1F,0xEF,0xFF,0x3F,0xF1,0xFF,0x1F,0xF8,0x1F,0xF0, \
	0x3F,0xE3,0xFE,0x7F,0xE7,0xFE,0x1F,0xEF,0xFF,0x3F,0xF1,0xFF,0x1F,0xF0,0x1F,0xF0, \
	0x3F,0xE3,0xFE,0x3F,0xE7,0xFE,0x1F,0xE7,0xFF,0x3F,0xF1,0xFF,0x1F,0xF0,0x1F,0xF8, \
	0x3F,0xE3,0xFE,0x3F,0xE7,0xFE,0x1F,0xE7,0xFF,0x3F,0xF1,0xFF,0x3F,0xF0,0x0F,0xF8, \
	0x3F,0xE3,0xFE,0x3F,0xE7,0xFE,0x1F,0xE7,0xFF,0x1F,0xF1,0xFF,0x3F,0xE0,0x0F,0xF8, \
	0x1F,0xF7,0xFE,0x3F,0xF7,0xFE,0x1F,0xE3,0xFF,0x1F,0xFB,0xFF,0x3F,0xE0,0x0F,0xFC, \
	0x1F,0xFF,0xFE,0x3F,0xFF,0xFC,0x1F,0xE3,0xFF,0x1F,0xFF,0xFF,0x7F,0xE0,0x07,0xFC, \
	0x1F,0xFF,0xFE,0x1F,0xFF,0xFC,0x1F,0xE3,0xFF,0x0F,0xFF,0xFF,0x7F,0xC0,0x07,0xFC, \
	0x0F,0xFF,0xFC,0x1F,0xFF,0xF8,0x1F,0xE3,0xFF,0x0F,0xFF,0xFF,0xFF,0xC0,0x07,0xFE, \
	0x07,0xFF,0xF8,0x0F,0xFF,0xF0,0x1F,0xE1,0xFF,0x07,0xFF,0xFF,0xFF,0xC0,0x07,0xFE, \
	0x03,0xFF,0xF0,0x03,0xFF,0xE0,0x1F,0xE1,0xFF,0x03,0xFF,0x3F,0x00,0x00,0x00,0x00, \
	0x00,0xFF,0xC0,0x00,0xFF,0x80,0x00,0x00,0x00,0x00,0xFE,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, \
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00
#endif

// i2c Display Module
#define DisplayName "Display"

// i2C OLED Display
class DisplayAddon : public GPAddon
{
public:
	virtual bool available();
	virtual void setup();
	virtual void preprocess() {}
	virtual void process();
	virtual std::string name() { return DisplayName; }
private:
    bool updateDisplayScreen();
	void drawStatusBar(Gamepad*);
	void initMenu(char**);
	bool pressedUp();
	bool pressedDown();
	bool pressedLeft();
	bool pressedRight();
	const DisplayOptions& getDisplayOptions();
	bool isDisplayPowerOff();
	void setDisplayPower(uint8_t status);
	uint32_t displaySaverTimeout = 0;
	int32_t displaySaverTimer;
	uint8_t displayIsPowerOn = 1;
	uint32_t prevMillis;
	std::string statusBar;
	Gamepad* gamepad;
	bool configMode;
	GPGFX* gpDisplay;
	GPScreen* gpScreen;
	DisplayMode currDisplayMode;
    DisplayMode prevDisplayMode;
	bool turnOffWhenSuspended;
};

#endif
